rust/otap-dataflow/docs/tracing-proposal.md

```diff
@@ -0,0 +1,145 @@
+# Proposal: OpenTelemetry-based Tracing for our dataflow engine
+
+## Problem Statement
+
+Modern observability pipelines are increasingly complex. In our system, these
+pipelines are defined as Directed Acyclic Graphs (DAGs) of interconnected
+processing nodes (receivers, processors, and exporters). In such a system,
+diagnosing data loss, performance bottlenecks, misrouted signals, stateful
+processing delays, or the impact of control-plane actions becomes challenging.
+
+**Traditional metrics and logs alone are insufficient** for root-cause analysis,
+end-to-end flow visibility, and operational insights, especially under dynamic
+configuration and sampling regimes.
+
+**Goal**:
+Leverage OpenTelemetry tracing primitives (traces, spans, events, links)
+to enable robust debugging, troubleshooting, monitoring, and optimization of the
+entire dataflow engine, **without imposing excessive overhead or losing critical
+information through sampling.**
+
+## System Overview
+
+The dataflow engine is structured as a DAG of the following node types:
+
+- **Receivers**: Ingest and translate data from external sources (may have
+  multiple output ports).
+- **Processors**: Transform, filter, batch, or enrich data (may have multiple
+  output ports).
+- **Exporters**: Deliver processed data to external systems.
+
+Each node features:
+
+- A **unique id** and **type** (i.e. urn).
+- A **control channel** for orchestration commands (e.g. config_update, ack,
+  timer_tick, shutdown).
+- A **pdata channel** (processors/exporters) for ingesting batches of signals (
+  metrics, logs, events, spans).
+
+**PData messages** are batches containing a single signal type.
+
+## Tracing Modes
+
+- **No Tracing**: No spans emitted.
+- **Tail Sampling**: Only selected traces (e.g. errors, random sample) are
```

Comment:
why only tail sampling? Wouldn't it be expensive (perf costs), to create/manage spans and drop it via tail sampling?

Reply:
I just listed modes that seemed useful to me, but this list is definitely open.

Regarding the efficiency of tail sampling in this specific context, I think we can implement it very efficiently because we control everything and are not operating in a general context.